### ISSUE TITLE
Kerberos auth doesn't work when the host is an alias

### ISSUE THREAD

**Issue by AndreaGiardini** [kennethreitz/requests#2502]
Hi

I'm using the request module in combination with request-kerberos. Since the server i'm querying is behind a load balancer it often happens that the authentication is done with host1 and the query on host2, leading to error 401 Unauthenticated.

I manage to workaround this problem this way:
```
host = socket.gethostbyaddr(random.choice(socket.gethostbyname_ex(host)[2]))[0]
```

Don't know if i should open this issue in `requests-kerberos` as well.

**Comment by Lukasa** [comment]
requests-kerberos should authenticate every connection so long as you attach it as an auth handler to the _session_, not to individual requests. Ideally that should avoid this problem.

**Comment by AndreaGiardini** [comment]
Do you mean something like this?

```
s = requests.Session()
s.auth = HTTPKerberosAuth()
s.get('http://my.beautiful.alias')
```

Because this example (on their project page) doesn't work in my case:

```
requests.get("http://my.beautiful.alias", auth=HTTPKerberosAuth())
```

**Comment by Lukasa** [comment]
That is exactly what I meant. =) Does that work?

**Comment by AndreaGiardini** [comment]
In my case it doesn't...
If it connects to the same server behind the load balancer it works, if it connects to a different one then i have authentication failed.

**Comment by Lukasa** [comment]
Interesting. Do all the endpoints share the same hostname?

**Comment by Lukasa** [comment]
The reason I ask is that Kerberos with loadbalancers is...[tricky](http://serverfault.com/questions/365065/kerberos-authentication-through-load-balancer).

**Comment by AndreaGiardini** [comment]
That's my case:

```
➜  host judy.my.com
judy.my.com is an alias for foremanlb.my.com.
foremanlb.my.com has address X
foremanlb.my.com has address Y
```

**Comment by Lukasa** [comment]
It would be interesting to see if any other kerberos client implementation functions here. I think Chrome has a Kerberos implementation...

**Comment by sigmavirus24** [comment]
So HTTPKerberosAuth requires the same IP each time for authentication to continue working? I'm not sure that's the problem of requests and I'm very sure we don't want to randomly choose a value to give the authenticator if in fact that's a potential solution.

**Comment by AndreaGiardini** [comment]
I didn't investigate deeply tbh...
Also requests need to pick up a random value and not the alias, otherwise it doesn't work.
This is the only way afaik:

```
host = socket.gethostbyaddr(random.choice(socket.gethostbyname_ex(host)[2]))[0]
requests.get(host, auth=HTTPKerberosAuth())
```

The thing is that both requests and HTTPKerberosAuth need to make their calls to the same server behind the alias.

**Comment by christian-korneck** [comment]
not sure if this might help, but I had a similar problem and it seems to have gone after setting SPNs for the alias names. In my case, I had a webserver bound to an Active Directory domain with a specific URL asking for Kerberos auth. Without SPNs for the alias name in place, auth didn't work. I ran this with DomainAdmin privileges to register the SPNs:

setspn /A host/myaliashostname realhostname
setspn /A http/myaliashostname realhostname

In my case, I also had to reboot the (Windows) client machine on which I'm running the python/requests client (had problems before reboot).
